tjpcov/nmt_tools.py

```diff
@@ -0,0 +1,438 @@
+
+    """
+    tr = sacc_data.get_tracer(tracer)
+    if tr.quantity in ['cmb_convergence', 'galaxy_density']:
+        return 1
+    elif tr.quantity == 'galaxy_shear':
+        return 2
+    else:
+        raise ValueError(f'tracer.quantity {tr.quantity} not implemented.')
+
+
+def get_tracer_spin(sacc_data, tracer):
+    """
+    Return the spin of a given tracer
+
+    Parameters:
+    -----------
+        sacc_data (Sacc):  Data Sacc instance
+        tracer (str):  Tracer name
+
+    Returns:
+    --------
+        spin (int):  Spin of the given tracer
+
+    """
+    tr = sacc_data.get_tracer(tracer)
+    if tr.quantity in ['cmb_convergence', 'galaxy_density']:
+        return 0
+    elif tr.quantity == 'galaxy_shear':
+        return 2
+
+
+def get_tracer_comb_spin(sacc_data, tracer_comb):
+    """
+    Return the spins of a pair of tracers
+
+    Parameters:
+    -----------
+        sacc_data (Sacc):  Data Sacc instance
+        tracer_comb (tuple):  List or tuple of a pair of tracer names
+
+    Returns:
+    --------
+        s1 (int):  Spin of the first tracer
+        s2 (int):  Spin of the second tracer
+
+    """
+    s1 = get_tracer_spin(sacc_data, tracer_comb[0])
+    s2 = get_tracer_spin(sacc_data, tracer_comb[1])
+
+    return s1, s2
+
+
+def get_tracer_comb_dof(sacc_data, tracer_comb):
+    """
+    Return the degrees of freedom of the Cell of a pair of tracers
+
+    Parameters:
+    -----------
+        sacc_data (Sacc):  Data Sacc instance
+        tracer_comb (tuple):  List or tuple of a pair of tracer names
+
+    Returns:
+    --------
+        dof (int):  Degrees of freedom of the Cell of the pair of tracers
+        given
+
+    """
+    dof1 = get_tracer_dof(sacc_data, tracer_comb[0])
+    dof2 = get_tracer_dof(sacc_data, tracer_comb[1])
+
+    return dof1 * dof2
+
+
+def get_datatypes_from_dof(dof):
+    """
+    Return the possible datatypes (cl_00, cl_0e, cl_0b, etc.) given a number
+    of degrees of freedom
+
+    Parameters:
+    -----------
+        dof (int):  Degrees of freedom of the Cell of the pair of tracers
+        given
+
+    Returns:
+    --------
+        datatypes (list):  List of data types assotiated to the given degrees
+        of freedom
+
+    """
+    # Copied from https://github.com/xC-ell/xCell/blob/069c42389f56dfff3a209eef4d05175707c98744/xcell/cls/to_sacc.py#L202-L212
+    if dof == 1:
+        cl_types = ['cl_00']
+    elif dof == 2:
+        cl_types = ['cl_0e', 'cl_0b']
+    elif dof == 4:
+        cl_types = ['cl_ee', 'cl_eb', 'cl_be', 'cl_bb']
+    else:
+        raise ValueError('dof does not match 1, 2, or 4.')
+
+    return cl_types
+
+
+def get_cl_for_cov(clab, nlab_cp, ma, mb, w):
+    """
+    Computes the coupled Cell that goes into the covariance matrix
+
+    Parameters:
+    -----------
+        clab (array): Fiducial Cell for the tracers a and b, used for in the
+        covariance estimation
+        nlab_cp (array): Coupled noise for the tracers a and b
```

Comment:
clab is the theoretical C_ell and goes from ell=0 to ell = 3 * nside - 1. 

The reason to pass the fiducial C_ell and the coupled noise is that these are the quantities computed  by TJPCov (following `cl_gaussian_cov` implementation).  Not sure this answer your question, though

Reply:
It is not clear to me what coupled noise means in this context. How is it different from simple noise power spectra, for example shot noise/shape noise, i.e. 1/n_g like terms?

Reply:
It is not different. `nlab_cp` is the simple noise power spectra `SN` that you can obtain with e.g. 1/n_g, as you say, with the correct shape that needs NaMaster (i.e. [ncls, nbpw]). The reason to call it "coupled" is that it is the noise that you subtract tto the coupled Cell (the one you compute from the maps). If you decouple the observed Cell, you need to subtract the "uncoupled noise" (i.e. the Nell multiplied by the inverse of the mode coupling matrix)

Reply:
I would argue that 1/n_g and the noise subtracted from pseudo-C_ell are not the same. In the simplest case they differ by f_sky factor. Typically when we write the covariance as, Cov ~ M* diag((C_ell+1/n_g)^2). We do not use the pseudo-C_ell noise. 
Also for the white noise, M^{-1} N_ell = 1/n_g, if you assume signal and noise window are the same. 

Perhaps I am missing some Namaster context here? It seems somewhat inconsistent to me to expect C_ell and coupled noise as user inputs. I think we should handle all window dependent calculations internally.

Reply:
I'm not sure I see your point. This function (`get_cl_for_cov`) is only called within https://github.com/LSSTDESC/TJPCov/blob/d471c37fe0797f2addfa113db3271a28f2411c54/tjpcov/main.py#L424 and the noise that is passed is the one that is computed somewhere else in the code. The user has no need to use anything in `nmt_tools`, so it is already handled internally. Also, note that, as you say, we do not use the pseudo-C_ell noise in `get_cl_for_cov`.

The reason to accept the noise as input is that, in this way, we can compute the `cl` that goes inside the covariance in just one line  https://github.com/LSSTDESC/TJPCov/blob/d471c37fe0797f2addfa113db3271a28f2411c54/tjpcov/main.py#L538-L539 Furthermore, I think `get_cl_for_cov` is a very general way of coding the operations that we need to do to obtain the `Cell` that goes into the covariance: it depends on the fiducial Cell, the noise, the masks of the fields and the (already computed) coupling matrix. The noise for galaxy clustering and shear can be obtained analytically but I don't think this is always the case (although the modelling or the loading of this possibly external noise would be done somewhere else in the code).

 I could remove the noise as input in the function but then I would have two options (that I can see quickly now):

* call `get_cl_for_cov` and add the noise contribution in the next line, which would involve dividing again by `np.mean(ma * mb)` or,
*  I would have to decouple the noise, add it to the fiducial Cell and then call `get_cl_for_cov`.

In both cases, the code will become more convoluted, with more `if np.mean(ma*mb) == 0`, etc.

Reply:
In case this helps: we always refer to "coupled" quantities as those corresponding to the masked field (i.e. \tilde{a} = w * a, where w is the mask). So the "coupled noise power spectrum" is the power spectrum of the noise component of a given field. As described in Andrina's paper, these (the coupled noise spectrum and the coupled signal spectrum) are the quantities you need to compute the covariance matrix. So Carlos' strategy here is (I assume, I haven't looked at the code):
- Get the coupled noise spectrum, whose calculation depends on the specifics of the fields being analysed (i.e. it's not simply 1/n when things are masked). I think it makes sense to expect this quantity to be computed by e.g. TXPipe.
- Get the theoretical signal power spectrum and then turn it into its coupled version internally by multiplying by the MCM.
- Then put everything together to get you the C_ell needed for the Gaussian covariance.

If a worry is that this interface is too complicated for someone who won't use this through TXPipe and just wants a quick covariance, in those cases users could just pass zero noise and a theory C_ell that includes their model for both signal and noise.

Reply:
> I'm not sure I see your point. This function (`get_cl_for_cov`) is only called within
> 
> https://github.com/LSSTDESC/TJPCov/blob/d471c37fe0797f2addfa113db3271a28f2411c54/tjpcov/main.py#L424
> 

Yes, the noise being passed from the main.py is shot noise, which is 1/n_g. It is not the coupled noise, i.e., w*1./n_g. No? 
My point is that for consistency and ease of understanding of the code, if we call "w.couple_cell(clab)" here, then we should do something like w.couple_Nell(Nlab) as well. I do not see any reason to do these at different places.

Reply:
Ok. Now I see. I thought `SN` was the coupled noise. If it is not, we can pass  `cl[13] + SN[13]` to `get_cl_for_cov` and couple the sum with NaMaster. However, I'm surprised. I thought that the coupled noise was the one you could compute analytically and was flat (e.g. Andrina's or NaMaster's paper), but not the uncoupled one, which you obtain later multiplying by the inverse of the mode coupling matrix ( M^1 * Nell_coupled). @damonge?

Reply:
I think the **coupled** noise spectrum (as opposed to the decoupled one) should be an input to this function, since this is something that depends entirely on the properties of the maps being correlated and that can be calculated from it. I would also advocate strongly against computing the coupled noise power spectrum ourselves by multiplying by the MCM. When there is significant mode-coupling (e.g. for cosmic shear), `np.dot(MCM, noise_decoupled)` will not in general be the same as `noise_coupled`, since `noise_coupled` is not band-limited (e.g. very often it's flat, and sometimes it's even blue). Furthermore, `noise_decoupled` is not generally something that can be computed from the maps directly, whereas `noise_coupled` is.

Reply:
Ok there are few things here:

1. To clarify, I thought we are talking about noise in pseudo-C_ell, so when I say un-convolved noise I simply mean 1/n_g like noise and not M^-1 N_ell coupled. As far as I know, we never need to compute M^-1 N_ell coupled during the analysis.
2. White noise stays white noise irrespective of what we convolve/deconvolve it with. A constant function convolved with any window results in a constant function. So convolved/deconvolved shot noise spectra only differ by a constant and we can compute that easily. The pseudo-C_ell shot noise is simply mean(window)*1/n_g. 
3. I agree that coupled noise spectrum depends on properties of maps, so does every other quantity we are computing. I do not quite follow why we should force the user to compute some of them and we compute the others separately. Calculation for a general coupled noise is same as that for coupled signal (it is simply M_N N_ell, where N_ell is the scale dependent unconvolved noise). In the context of covariance calculations, it is not very expensive computationally. 

I think it is good if you want to have an option to allow user to input coupled noise spectra. But it needs to be a separate argument to TJPcov, something like SN_coupled and the code should raise an error when user passes the wrong input (i.e. SN when we expect SN_coupled or vice versa). Or if SN_coupled=None, we raise a warning and compute it ourselves when needed.

Reply:
I cannot see how M * N_ell or M^-1 * N_ell_coupled can be constant unless M is proportional to the identity matrix (which will not be the case in general). For instance, see the plot attached (or Fig. 3 of Andrina's paper https://arxiv.org/pdf/2010.09717.pdf). Even in the simpler case of clustering you can see a bit of mode coupling at very large ell (see 2nd and 3rd plots attached), although, in this case, it might come from some edge effects (but these will also be present in the signal Cell).

I think @damonge is not advocating for forcing the user to compute the coupled noise, but saying that the quantity that we should compute internally is the coupled noise (which we can write analytically), instead of the uncoupled (which, in general, we cannot write analytically).

![image](https://user-images.githubusercontent.com/17812754/130029123-0bcebfb1-445b-4f16-81b1-d1a0370dbec0.png)
![image](https://user-images.githubusercontent.com/17812754/130029902-f5ddc9dd-a013-4335-ba3e-282e4b422bc4.png)
![image](https://user-images.githubusercontent.com/17812754/130031930-813bd611-d0e4-49db-bee9-2295fd8e129a.png)

Reply:
> I cannot see how M * N_ell or M^-1 * N_ell_coupled can be constant (given that N_ell or N_ell_coupled are constant)

You can see this directly from the equation for pseudo-C_ell noise (for example eq. 2.32 in Andrina's paper you linked) where M_N N_ell = mean(window)*1./n_g is a constant. 

The scale dependence you are seeing is the consequence of scale cuts (or edge effects) imposed on M (assuming input noise power is constant). In principle M needs to be computed out to ell of infinity to get the proper couplings. For signal, we are lucky that C_ell ~ 1/ell and ell cuts on M work out to be fine in most cases. For the white noise case, it doesn't work and we either need to use analytical forms or increase the ell range by a factor of few in M. I briefly discuss this in the beginning of section 3 in https://arxiv.org/pdf/2105.04548.pdf. 

>unless M is proportional to the identity matrix (which will not be the case in general). For instance, see the plot attached (or Fig. 3 of Andrina's paper https://arxiv.org/pdf/2010.09717.pdf). Even in the simpler case of clustering you can see a bit of mode coupling at very large ell (see 2nd and 3rd plots attached), although, in this case, it might come from some edge effects (but these will also be present in the signal Cell).
> 
> I think @damonge is not advocating for forcing the user to compute the coupled noise, but saying that the quantity that we should compute internally is the coupled noise (which we can write analytically), instead of the uncoupled (which, in general, we cannot write analytically).

Ok, if you want to compute it internally within nmt_tools, that is fine. My impression was that you guys want it to be user input.

Reply:
It seems to me the easiest way forward would be to accept both SN and SN_coupled. If SN_coupled is passed, we use it directly, otherwise, we take SN and we couple it internally (in my opinion this is not desirable, since due to the band effects we've mentioned above, the resulting coupled noise will be unreliable).

Does this sound like a good compromise?

Reply:
At the moment, SN is computed internally from given `n_g` and `sigma_e` for the lens and source tracers in https://github.com/LSSTDESC/TJPCov/blob/a0c23af6c9d9ee462427aca1eb71b94a21db3d42/tjpcov/main.py#L382 and https://github.com/LSSTDESC/TJPCov/blob/a0c23af6c9d9ee462427aca1eb71b94a21db3d42/tjpcov/main.py#L387. 

In order to keep this PR as related to the NaMaster implementation as possible, what would you think of adding a option tracer_Noise_coupled in this function? https://github.com/LSSTDESC/TJPCov/blob/d471c37fe0797f2addfa113db3271a28f2411c54/tjpcov/main.py#L424-L426
If it is given, `get_cl_for_cov` will use it as is now in https://github.com/LSSTDESC/TJPCov/blob/d471c37fe0797f2addfa113db3271a28f2411c54/tjpcov/nmt_tools.py#L123 
otherwise, it will be added to `clab` before coupling.

We can open a issue about generalizing this in the code and add a method to compute the analytical noise in a later PR.

Reply:
@carlosggarcia The proposal sounds good. Thanks!

I would recommend not adding noise to C_ell for coupling. Like @damonge said, it can leads to unreliable results. Why not simply use the relation: N_ell_coupled= N_ell * np.mean(mask)? I think we can assume that the noise passed will be shot noise like in most cases. If you want to be extra careful, you can also print a warning once when doing this computation so that the user knows.

Reply:
I would add that warning, yes, since this is only correct for flat spectra. But this sounds like a good idea

Reply:
@sukhdeep2 done in https://github.com/LSSTDESC/TJPCov/pull/21/commits/2249449d9f5b614650df5efad36af6f952a8e0cb! Thanks!

Reply:
Looks good to me. Thanks!